### DISCUSSION TITLE
Jupyter workspace refuses to save in the workspace when running PL locally, fixed by changing contents of Dockerfile and rebuilding

### DISCUSSION THREAD

**Discussion by justindr21** [PrairieLearn/PrairieLearn#6268] [General]
Thanks to Eric Huber helped me with this whole grading ordeal. In short, I was able to create grading images for jupyter workspace questions but I could not save the workspace itself, causing the grader to grade nothing all the time since it can never be saved. Apparently it has something to do with which user is running PL and that wasn't giving me any permission to make changes or save things.

I asked if this problem would only happen to me and the answer was "not really" so in case anyone runs into this problem too, here you go!

Eric suggested that I rebuild docker by creating a new dockerfile under the name "pltest.Docker". I just copied the original dockerfile that should just be named "Dockerfile" in the PL codebase, pasted it in the same directory and then renamed it. 

You can either paste the following block of code at the bottom of the copy:

```
RUN printf '{\n\
    "workspaceJobsDirectoryOwnerUid": 1001,\n\
    "workspaceJobsDirectoryOwnerGid": 1001\n\
}\n\
' > /PrairieLearn/config.json && cat /PrairieLearn/config.json

```

or you could replace the entire contents of the file with the following block of code:

```
FROM prairielearn/prairielearn

RUN printf '{\n\
    "workspaceJobsDirectoryOwnerUid": 1001,\n\
    "workspaceJobsDirectoryOwnerGid": 1001\n\
}\n\
' > /PrairieLearn/config.json && cat /PrairieLearn/config.json

```
After that, run the following in the terminal of your choice, I was using WSL using Ubuntu

`docker build -t pltest -f pltest.Dockerfile .`

where "pltest.Docker" is the name of the new dockerfile that you just created. (I say this in case you decided to rename the copy something else)

We can let that build and then call the regular docker command with pltest at the bottom of the command instead of prairielearn/prairielearn, so something like this (linux):

```
docker run -it --rm -p 3000:3000 \
  -v "$PWD":/course \
  -v "$HOME/pl_ag_jobs:/jobs" \
  -e HOST_JOBS_DIR="$HOME/pl_ag_jobs" \
  -v /var/run/docker.sock:/var/run/docker.sock \
  --add-host=host.docker.internal:172.17.0.1 \
  pltest
```

And that solved my issue at least. 

So if you are facing an issue where you cannot save files in a jupyter workspace or you get an error that says something like "no permission" when you try to create a new file in the workspace (which you should be able to do normally), then this might fix it. 

Thanks again to Eric Huber.

**Comment by echuber2** [comment]
Here is another post documenting how to run as 1001:1001 with an extra setting on the Docker invocation (avoiding the need to rebuild an image): https://github.com/PrairieLearn/PrairieLearn/discussions/8897
